upload-api/access-types.ts

```diff
@@ -0,0 +1,69 @@
+import * as Ucanto from '@ucanto/interface'
+import { ProviderAdd } from '@web3-storage/capabilities/src/types'
+
+interface ByAudience {
+  audience: Ucanto.DID<'key' | 'mailto'>
+}
+export type Query = ByAudience
+
+
+/**
+ * TODO use the version in w3up access-api/src/types/delegations.ts
+ */
+export interface DelegationsStorage<
+  Cap extends Ucanto.Capability = Ucanto.Capability
+> {
+  /**
+   * write several items into storage
+   *
+   * @param delegations - delegations to store
+   */
+  putMany: (
+    ...delegations: Array<Ucanto.Delegation<Ucanto.Tuple<Cap>>>
+  ) => Promise<unknown>
+
+  /**
+   * get number of stored items
+   */
+  count: () => Promise<bigint>
+
+  /**
+   * find all items that match the query
+   */
+  find: (query: Query) => AsyncIterable<Ucanto.Delegation<Ucanto.Tuple<Cap>>>
```

Comment:
yep, works for me!

Reply:
Hm - after looking at this a bit more I'm second guessing this - we can't really know if the underlying system will support streaming results and it does seem valuable to allow novel implementations to stream results back from this function if they want. 

Since the system is already implemented with this async implementation I'm definitely inclined to keep it as-is to both simplify this change and give future implementors maximum flexibility.

@Gozala what do you think? I'm still open to making this a regular `Iterable` but am inclined to keep it as is...

Reply:
I'm of a position that this is premature optimization. If we deal with a system that allows us to do streaming we:

1. Should still consider if that would be a right choice, or e.g. whether paging would serve us better.
2. If we want to take advantage of streamed output then we can design solution that fits best the use case, which may or may not be async iterables.
    - For what it's worth in my experience async iterable are rarely the right tool for the job. Even with async streams you often want to have some control of coordination (telling sender to pause / slow down or speed up) and async iterables unlike e.g. web streams don't provide that. They also provide little insight if multiple readers would get same output or they'll get random parts when reading concurrently.

To summarize I think `AsyncIterable`s provide no value now and likely aren't going to be a good choice in the future. That said leaving it out of scope for this PR is reasonable and I'm not going to object to it if that is what you want to do.

Reply:
ok I'm convinced! have changed this interface in https://github.com/web3-storage/w3up/pull/790/commits/026ab8726b73d0ce907a409898325f365bd3cab6